flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java

```diff
@@ -53,175 +58,260 @@
+			final BucketID bucketId,
+			final Path bucketPath,
+			final long initialPartCounter,
+			final PartFileWriter.PartFileFactory<IN, BucketID> partFileFactory,
+			final RollingPolicy<IN, BucketID> rollingPolicy) {
 
 		this.fsWriter = Preconditions.checkNotNull(fsWriter);
 		this.subtaskIndex = subtaskIndex;
 		this.bucketId = Preconditions.checkNotNull(bucketId);
 		this.bucketPath = Preconditions.checkNotNull(bucketPath);
 		this.partCounter = initialPartCounter;
+		this.rollingPolicy = Preconditions.checkNotNull(rollingPolicy);
 		this.partFileFactory = Preconditions.checkNotNull(partFileFactory);
 
-		this.pending = new ArrayList<>();
+		this.pendingPartsForCurrentCheckpoint = new ArrayList<>();
 	}
 
 	/**
-	 * Gets the information available for the currently
-	 * open part file, i.e. the one we are currently writing to.
-	 *
-	 * <p>This will be null if there is no currently open part file. This
-	 * is the case when we have a new, just created bucket or a bucket
-	 * that has not received any data after the closing of its previously
-	 * open in-progress file due to the specified rolling policy.
-	 *
-	 * @return The information about the currently in-progress part file
-	 * or {@code null} if there is no open part file.
+	 * Constructor to restore a bucket from checkpointed state.
 	 */
-	public PartFileInfo<BucketID> getInProgressPartInfo() {
-		return currentPart;
+	private Bucket(
+			final RecoverableWriter fsWriter,
+			final int subtaskIndex,
+			final long initialPartCounter,
+			final PartFileWriter.PartFileFactory<IN, BucketID> partFileFactory,
+			final RollingPolicy<IN, BucketID> rollingPolicy,
+			final BucketState<BucketID> bucketState) throws IOException {
+
+		this(
+				fsWriter,
+				subtaskIndex,
+				bucketState.getBucketId(),
+				bucketState.getBucketPath(),
+				initialPartCounter,
+				partFileFactory,
+				rollingPolicy);
+
+		restoreInProgressFile(bucketState);
+		commitRecoveredPendingFiles(bucketState);
+	}
+
+	private void restoreInProgressFile(final BucketState<BucketID> state) throws IOException {
+		if (state.hasInProgressResumableFile()) {
+			final RecoverableWriter.ResumeRecoverable resumable = state.getInProgressResumableFile();
+			inProgressPart = partFileFactory.resumeFrom(
+					bucketId, fsWriter, resumable, state.getInProgressFileCreationTime());
+		}
+	}
+
+	private void commitRecoveredPendingFiles(final BucketState<BucketID> state) throws IOException {
+		for (List<RecoverableWriter.CommitRecoverable> committables: state.getCommittableFilesPerCheckpoint().values()) {
+			for (RecoverableWriter.CommitRecoverable committable: committables) {
+				fsWriter.recoverForCommit(committable).commitAfterRecovery();
+			}
+		}
+	}
+
+	void merge(final Bucket<IN, BucketID> bucket) throws IOException {
+		Preconditions.checkNotNull(bucket);
+		Preconditions.checkState(Objects.equals(bucket.bucketPath, bucketPath));
+
+		// there should be no pending files in the "to-merge" states.
```

Comment:
This isn't relevant to this PR, but these comments would be much more valuable if a reasoning is included.